### PR TITLE
Lower tile to Krnl

### PR DESCRIPTION
**Pull request by chentong319**
The Tile Op in onnx is the same as bumpy tile without broadcast. That means the size of the repeats tensor is the same as the rank of the input and output tensor.
There are two ways to lower Tile to Krnl:

1. Default implementation
loop over output (I1, I2, ..,Ik, ...).
Loop body:
output[...,Ik, ...] = input[..., Ik mod input.shape[k], ...]

2. Alternative implementation
loop over input (I1, I2, ..., Ik, ...)  and repeats ( J0: 0 to  repeats[0], J1  0 to repeats[1], ..., Jk:0 to repeats[k], ...)
Loop body:
output[..., Ik+Jk*input.shape[k], ...] = input[..., Ik, ....]

I implemented both in Tile.cpp but choose the first one as the default. Unclear to me which one is better for optimization.

The alternative implementation passed the backend test.

### CODE REVIEW

**Comment by tjingrant** [comment]
Hi @chentong319 could you enable relevant node-level unit tests here:
https://github.com/onnx/onnx-mlir/blob/master/test/backend/test.py

**Comment by tjingrant** [comment]
@chentong319 unit test for tile operation seems to be failing, could you double-check the implementation?

**Comment by chentong319** [comment]
@tjingrant check-onnx-backend passed. There were two errors:
1. Should use input.shape[k] for mod or multiple, instead of repeats[k]
2. Can not use affineload when the subscript expression contains mod

File: src/Conversion/ONNXToKrnl/Tensor/Tile.cpp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+      auto loopVarVal = iterationBlock.getArguments()[i];
+      auto exprVal =
+          rewriter.create<UnsignedRemIOp>(loc, loopVarVal, inputDimSizeVal);
+      inputMemRefVal.emplace_back(exprVal);
+    }
+
+    // Load the value from input
+    // Tried to use affine load when the input has constant shape
+    // But got runtime complaint, perhaps due ot RemIOp
+    auto inputVal = rewriter.create<LoadOp>(loc, input, inputMemRefVal);
```
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToKrnl/Tensor/Tile.cpp:123]: mod is ok for affine maps:

affine-expr ::= `(` affine-expr `)`
              | affine-expr `+` affine-expr
              | affine-expr `-` affine-expr
              | `-`? integer-literal `*` affine-expr
              | affine-expr `ceildiv` integer-literal
              | affine-expr `floordiv` integer-literal
              | affine-expr `mod` integer-literal
              | `-`affine-expr
              | bare-id
              | `-`? integer-literal

The MLIR way would be to generate a map when the TILED DIMENSIONS are constant sized.

val = affine.load(input, output indices, affine_map<(o0, o1, o2) -> (o0, o1 mod tile1, o2 mod tile2)>)

You would need a memref_cast

%4 = memref_cast %1 : memref<12x4xf32, offset:?, strides: [?, ?]> to
                      memref<12x4xf32, offset:5, strides: [4, 1]>

%4 = memref_cast %1 : memref<S0, S1, S2> to memref<S0, S1 * tile1, S2 * tile2, affine_map<d0, d1 mod tile1, d2 mod tile2>>

Could you try if this work? And if it does not, at the very least we should ask the MLIR folks why it should not work...

File: src/Conversion/ONNXToKrnl/Tensor/Tile.cpp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    // Then store the value in the output.
+    rewriter.create<AffineStoreOp>(loc, inputVal, alloc, outputMemRefVal);
+
+    rewriter.replaceOp(op, alloc);
+
+    return success();
+  }
+};
+
+// This is the alternative way of lowering.
```
[the remark this answers is not in the document]
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/Tile.cpp:136]: I tested the alternative implementation.

File: src/Conversion/ONNXToKrnl/Tensor/Tile.cpp [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    // This implementation is to iterate the output tensor.
+    // The store has simple affine subscript expression.
+    // Alternative implementation is to iterate the input tensor and repeats.
+    // The load of elements in input tensor can be reused explicitly.
+    // But the subscript of store is not contigous, or even not affine.
+    // Alternative implementation can be found at the end of this file.
+    SmallVector<Value, 4> inputMemRefVal;
+    for (int i = 0; i < outputRank; ++i) {
+      Value inputDimSizeVal;
+      if (inputShape[i] == -1)
```
[the remark this answers is not in the document]
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/Tile.cpp:109]: Changed to dim Op

File: src/Conversion/ONNXToKrnl/Tensor/Tile.cpp (hunk 1 above)
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/Tile.cpp:123]: The problem is caused by not using AffineExpr, not the operation.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, thanks for the changes

**Review by AlexandreEichenberger — CHANGES_REQUESTED** [review]
Please do not use sea variable names directly in the mlir literal tests.
Please use the ```[[VAR:%.+]]``` instead.

**Review by AlexandreEichenberger — COMMENTED** [review]
@chentong319 ,

if the tiling is dynamic sizes (the amount of repeat is not constant), I believe you have

`j mod k` where `j` and `k` are variables, and that is not affine (affine allows `mod integer` only.

I am ok if we assert as of now... but this does not to work in general.

**Comment by chentong319** [comment]
@AlexandreEichenberger this operation is affine because it is `kth iteration variable % repeats[k]` and k is a constant and repeats[k] is a constant value for the loop. E.g. for loop (j, k), we have `j  % repeats[0]` and `k % repeats[1]`

**Review by AlexandreEichenberger — APPROVED** [review]
I get it, make sense. LGTM now.

**Review by AlexandreEichenberger — CHANGES_REQUESTED** [review]
Actually, you use 
```
affine map index % array size
```

if I look at the affine definition:

```
affine-expr ::= `(` affine-expr `)`
           ...
              | affine-expr `mod` integer-literal
```

but I think `dim` is more like a symbol

```
semi-affine-expr ::= `(` semi-affine-expr `)`
  ...
                   | semi-affine-expr `mod` symbol-or-const
                   | bare-id
                   | `-`? integer-literal

symbol-or-const ::= `-`? integer-literal | symbol-id
```
and the def for affine load is specific, it refer to an affine expression, not a semi-affine expression.

```
The “affine.load” op reads an element from a memref, where the index for each memref dimension is an affine expression of loop induction variables and symbols. The output of ‘affine.load’ is a new value with the same type as the elements of the memref. An affine expression of loop IVs and symbols must be specified for each dimension of the memref. The keyword ‘symbol’ can be used to indicate SSA identifiers which are symbolic.
```

So I don't think that we can use affine load in the current example. Only when the size of the input array are constant, and we use the integer value directly.

**Comment by doru1004** [comment]
We must make sure that the operation the map performs is indeed affine (according to the spec) otherwise we may end up generating the wrong code further down the line.

@AlexandreEichenberger thanks for showcasing this in your comment.

**Review by AlexandreEichenberger — COMMENTED** [review]
I took the output of your first test:
```
(onnx-mlir) alexe@pancetta maps % onnx-mlir-opt tile-test-tong2.mlir --lower-affine
tile-test-tong2.mlir:15:12: error: semi-affine expressions (modulo by non-const) are not supported
      %4 = affine.apply #map0(%arg1)[%3]
           ^
tile-test-tong2.mlir:18:12: error: semi-affine expressions (modulo by non-const) are not supported
      %6 = affine.apply #map0(%arg2)[%5]
           ^
#map0 = affine_map<(d0)[s0] -> (d0 mod s0)>
#map1 = affine_map<() -> (0)>
#map2 = affine_map<() -> (12)>
#map3 = affine_map<() -> (16)>


module {
  func @test_tile1(%arg0: memref<4x8xf32>) -> memref<12x16xf32> {
    %0 = alloc() : memref<12x16xf32>
    %1 = "krnl.global"() {name = "constant_0", shape = [2], value = dense<[3, 2]> : tensor<2xi64>} : () -> memref<2xi64>
    %2:2 = krnl.define_loops 2
    krnl.iterate(%2#0, %2#1) with (%2#0 -> %arg1 = 0 to 12, %2#1 -> %arg2 = 0 to 16) {
      %c0 = constant 0 : index
      %3 = dim %arg0, %c0 : memref<4x8xf32>
      %4 = affine.apply #map0(%arg1)[%3]
      %c1 = constant 1 : index
      %5 = dim %arg0, %c1 : memref<4x8xf32>
      %6 = affine.apply #map0(%arg2)[%5]
      %7 = load %arg0[%4, %6] : memref<4x8xf32>
      store %7, %0[%arg1, %arg2] : memref<12x16xf32>
    }
    return %0 : memref<12x16xf32>
  }
}
```

and fed it to `onnx-mlir-opt --lower-affine`, I got errors (apologies if this is not the right operation to call it to). Errors are listed here:

```
tile-test-tong2.mlir:15:12: error: semi-affine expressions (modulo by non-const) are not supported
      %4 = affine.apply #map0(%arg1)[%3]
           ^
tile-test-tong2.mlir:18:12: error: semi-affine expressions (modulo by non-const) are not supported
      %6 = affine.apply #map0(%arg2)[%5]
```

Interestingly, if I switch the `dim` by `const` I still get errors:

```
      %c4 = constant 4 : index
      %3 = affine.apply #map0(%arg1)[%c4]
```
which  means that even if we do constant propagation of dim of a constant memref, it will still not work.

Only when I use the integer literal inside of the map, can onnx-mlir-opt lower the code.

```
#map04 = affine_map<(d0) -> (d0 mod 4)>
#map08 = affine_map<(d0) -> (d0 mod 8)>

module {
  func @test_tile1(%arg0: memref<4x8xf32>) -> memref<12x16xf32> {
    %0 = alloc() : memref<12x16xf32>
    %1 = "krnl.global"() {name = "constant_0", shape = [2], value = dense<[3, 2]> : tensor<2xi64>} : () -> memref<2xi64>
    %2:2 = krnl.define_loops 2
    krnl.iterate(%2#0, %2#1) with (%2#0 -> %arg1 = 0 to 12, %2#1 -> %arg2 = 0 to 16) {
      %4 = affine.apply #map04(%arg1)
      %6 = affine.apply #map08(%arg2)
      %7 = affine.load %arg0[%4, %6] : memref<4x8xf32>
      affine.store %7, %0[%arg1, %arg2] : memref<12x16xf32>
    }
    return %0 : memref<12x16xf32>
  }
}
```

**Review by AlexandreEichenberger — COMMENTED** [review]
After careful considerations, I believe it would be best to directly use the AffineExpr computations inside of the affine.load as below:

```
module {
  func @test_tile1(%arg0: memref<4x8xf32>) -> memref<12x16xf32> {
    %0 = alloc() : memref<12x16xf32>
    %1 = "krnl.global"() {name = "constant_0", shape = [2], value = dense<[3, 2]> : tensor<2xi64>} : () -> memref<2xi64>
    %2:2 = krnl.define_loops 2
    krnl.iterate(%2#0, %2#1) with (%2#0 -> %arg1 = 0 to 12, %2#1 -> %arg2 = 0 to 16) {
      %7 = affine.load %arg0[%arg1 mod 4, %arg2 mod 8] : memref<4x8xf32>
      affine.store %7, %0[%arg1, %arg2] : memref<12x16xf32>
    }
    return %0 : memref<12x16xf32>
  }
}
```

and it works. 

From my dining into mlir, there is an example in mlir/lib/Transform/Utils/LoopUtils.cpp of a create<AffineLoadOp> that uses affine map.

**Comment by chentong319** [comment]
This issue is caused by the mod operation (similar situation for div and multiple).  Only + and - allow both operands to be AffineExpr.
Will restore to previous version of code.

**Comment by AlexandreEichenberger** [comment]
Tong, that is not my suggestion. I suggest that we generate the previous version with dyn sizes, or the one below with constant sizes.

```
      %7 = affine.load %arg0[%arg1 mod 4, %arg2 mod 8] : memref<4x8xf32>
```
mods are allowed with integers.

**Comment by tungld** [comment]
> (onnx-mlir) alexe@pancetta maps % onnx-mlir-opt tile-test-tong2.mlir --lower-affine
tile-test-tong2.mlir:15:12: error: semi-affine expressions (modulo by non-const) are not supported
      %4 = affine.apply #map0(%arg1)[%3]
           ^
tile-test-tong2.mlir:18:12: error: semi-affine expressions (modulo by non-const) are not supported
      %6 = affine.apply #map0(%arg2)[%5]

Canonicalize before `--lower-affine` would help: onnx-mlir-opt tile-test-tong2.mlir --canonicalize --lower-affine
```mlir
#map0 = affine_map<() -> (0)>
#map1 = affine_map<() -> (12)>
#map2 = affine_map<() -> (16)>


module {
  func @test_tile1(%arg0: memref<4x8xf32>) -> memref<12x16xf32> {
    %0 = alloc() : memref<12x16xf32>
    %1 = "krnl.global"() {name = "constant_0", shape = [2], value = dense<[3, 2]> : tensor<2xi64>} : () -> memref<2xi64>
    %2:2 = krnl.define_loops 2
    krnl.iterate(%2#0, %2#1) with (%2#0 -> %arg1 = 0 to 12, %2#1 -> %arg2 = 0 to 16) {
      %c4 = constant 4 : index
      %3 = remi_signed %arg1, %c4 : index
      %c0 = constant 0 : index
      %4 = cmpi "slt", %3, %c0 : index
      %5 = addi %3, %c4 : index
      %6 = select %4, %5, %3 : index
      %c8 = constant 8 : index
      %7 = remi_signed %arg2, %c8 : index
      %c0_0 = constant 0 : index
      %8 = cmpi "slt", %7, %c0_0 : index
      %9 = addi %7, %c8 : index
      %10 = select %8, %9, %7 : index
      %11 = load %arg0[%6, %10] : memref<4x8xf32>
      store %11, %0[%arg1, %arg2] : memref<12x16xf32>
    }
    return %0 : memref<12x16xf32>
  }
}
```

**Comment by AlexandreEichenberger** [comment]
@tungld Thanks, good to know. But presumably that worked only in the example where the dims were constants, correct?

**Comment by chentong319** [comment]
@AlexandreEichenberger We are on the same page: there is a previous version that choose affine.load only when shape is constant. 

As @tungld pointed out, the --canonicalize  flag can help. Also, the current version passed the backend test. That's a sign that we can rely on the system to handle DimOp for constant and constant propagation through map. Thus, I kept the most of the code and just added a check of the shape to select affine load or ordinary load.

**Comment by AlexandreEichenberger** [comment]
@chentong319 

as long as the version with the const dims can use the affine maps, affine.apply and that the one with the dyn dimensions use the compute directly.

I tested Tung's additional canonnalize option, and it simplifies the dims of constant sizes fine, good to know. But when adding a `?` in the memref, it chokes again on the non-affine.


``` mlir
(onnx-mlir) alexe@pancetta maps % onnx-mlir-opt tile-test-tong2-dyn.mlir --canonicalize --lower-affine
tile-test-tong2-dyn.mlir:19:12: error: semi-affine expressions (modulo by non-const) are not supported
      %7 = affine.load %arg0[%4, %6] : memref<?x8xf32>
           ^
#map0 = affine_map<()[s0, s1, s2] -> (s1 mod s0, s2 mod 8)>
#map1 = affine_map<() -> (0)>
#map2 = affine_map<() -> (12)>
#map3 = affine_map<() -> (16)>


module {
  func @test_tile1(%arg0: memref<?x8xf32>) -> memref<12x16xf32> {
    %c0 = constant 0 : index
    %0 = alloc() : memref<12x16xf32>
    %1 = "krnl.global"() {name = "constant_0", shape = [2], value = dense<[3, 2]> : tensor<2xi64>} : () -> memref<2xi64>
    %2:2 = krnl.define_loops 2
    krnl.iterate(%2#0, %2#1) with (%2#0 -> %arg1 = 0 to 12, %2#1 -> %arg2 = 0 to 16) {
      %3 = dim %arg0, %c0 : memref<?x8xf32>
      %4 = affine.load %arg0[symbol(%arg1) mod symbol(%3), symbol(%arg2) mod 8] : memref<?x8xf32>
      store %4, %0[%arg1, %arg2] : memref<12x16xf32>
    }
    return %0 : memref<12x16xf32>
  }
}
````

**Comment by AlexandreEichenberger** [comment]
Interestingly (for me at least), the cannonalize moves the apply you added directly as operations inside of the affine.load.

I checked also that if you change the `affine.load` to the `std.load`, your code works as is... So double check too, but it looks like the only change you need to do is to determine `affine` vs `std`.

**Review by AlexandreEichenberger — APPROVED** [review]
Once you have the changes, it should be all good. But check that the cannonalize phase is used too in the right place. Thanks

**Comment by chentong319** [comment]
It is neat to handle the constant dimension and unknown dimension in this way. I think this is the reason that MLIR has the semi-affine extension over the restrict affine.

Thanks @AlexandreEichenberger  and @tungld. Please review my PR for flatten, which is coming soon.

**Comment by chentong319** [comment]
@jenkins-droid test this please.

**Pull request merged**